core/blockchain/impl/level_db_block_tree.cpp

```diff
@@ -0,0 +1,269 @@
+        // should not be here: any node in our tree must be a descendant of the
+        // last finalized block
+        return Error::INTERNAL_ERROR;
+      }
+      node = *node.get().parent;
+    }
+
+    return result;
+  }
+
+  LevelDbBlockTree::BlockHashVecRes LevelDbBlockTree::longestPath() {
+    return getChainByBlock(deepestLeaf());
+  }
+
+  const primitives::BlockHash &LevelDbBlockTree::deepestLeaf() const {
+    return tree_meta_.deepest_leaf.block_hash;
+  }
+
+  std::vector<primitives::BlockHash> LevelDbBlockTree::getLeaves() const {
+    std::vector<primitives::BlockHash> result;
+    result.reserve(tree_meta_.leaves.size());
+    std::transform(tree_meta_.leaves.begin(), tree_meta_.leaves.end(),
+                   result.begin(),
+                   [](const auto *node) { return node->block_hash; });
+    return result;
+  }
+
+  LevelDbBlockTree::BlockHashVecRes LevelDbBlockTree::getChildren(
+      const primitives::BlockHash &block) {
+    OUTCOME_TRY(node_wrapped, tree_.getByHash(block));
+    auto &node = node_wrapped.get();
+
+    std::vector<primitives::BlockHash> result;
+    result.reserve(node.children.size());
+    for (const auto &child : node.children) {
+      result.push_back(child.block_hash);
+    }
+
+    return result;
+  }
+
+  primitives::BlockHash LevelDbBlockTree::getLastFinalized() const {
+    return tree_meta_.last_finalized.block_hash;
+  }
+
+  void LevelDbBlockTree::prune() {
+    // remove all non-finalized blocks from both database and meta
+    std::vector<std::pair<primitives::BlockHash, primitives::BlockNumber>>
+        to_remove;
+
+    auto last_node_hash = tree_meta_.last_finalized.block_hash;
+    auto current_node = std::ref(*tree_meta_.last_finalized.parent);
+    do {
+      auto &node_children = current_node.get().children;
+
+      // memorize the hashes and numbers of block to be removed
+      for (const auto &child : node_children) {
+        if (child.block_hash != last_node_hash) {
+          to_remove.emplace_back(child.block_hash, child.depth);
+        }
+      }
+
+      // remove them from the meta
+      node_children.erase(
+          std::remove_if(node_children.begin(), node_children.end(),
+                         [&last_node_hash](const auto &node) {
+                           return node.block_hash != last_node_hash;
+                         }),
+          node_children.end());
+
+      // go up to the next node
+      if (auto parent = current_node.get().parent) {
+        current_node = *parent;
+      }
+      last_node_hash = current_node.get().block_hash;
+    } while (!current_node.get().finalized);
+
+    // now, remove the blocks we remembered from the database
+    for (const auto &[hash, number] : to_remove) {
+      auto block_lookup_key = numberAndHashToLookupKey(number, hash);
+
+      auto header_lookup_key = prependPrefix(block_lookup_key, Prefix::HEADER);
+      (void)db_.remove(header_lookup_key);  // not really interested
```

Comment:
why? what if db_ for example was disconnected. How we can discover that? At very least we should add logs here